### ISSUE TITLE
glfwSetWindowOcclusionCallback to save CPU

### ISSUE THREAD

**Issue by xanather** [glfw/glfw#1828]
Throwback to PR https://github.com/glfw/glfw/pull/1123.

I know that MacOS supports properly obeying vsync now while the window is occluded; the scope of this issue is different.

There should be an API which is supported on all major platforms to determine whether we should even bother drawing and maybe sleeping the game loop for short periods if a window is occluded (different flag compared to iconified). Occlusion could mean: the computer is locked, no monitor attached, UAC takes over the screen, etc...

On my Direct3D backend I can use DXGI_STATUS_MODE_CHANGED and DXGI_PRESENT_TEST on IDXGISwapChain::Present() to determine when to slow down and not draw. Not sure if there is an applicable API for OpenGL/Vulkan backends...

**Comment by tombsar** [comment]
I don't believe this is provided by OpenGL or Vulkan APIs - probably need to query the window manager/compositor on X11.

**Comment by linkmauve** [comment]
On Wayland you naturally get throttled by the compositor if it doesn’t need your buffer at the moment, it will only send you a frame callback (requesting you to draw something) when it is going to display your window, and EGL uses the frame callback to implement `eglSwapBuffers()` waiting when `eglSwapInterval()` is set to 1 or above. If not, you will have to listen to frame callbacks yourself to know when to schedule a redraw or not.
